File: ios/App/App/Plugins/OfflineTilesPlugin/TileDownloadManager.swift
import Foundation

/// Downloads map tiles for a bounding box across a range of zoom levels and
/// stores them via `TileCacheManager`.
///
/// Uses slippy-map math to convert lat/lon bbox to tile coordinates, then
/// fetches vector, satellite, and terrain tiles concurrently (max 4 at a time).
final class TileDownloadManager {

    // MARK: - Types

    /// Progress callback delivered on every tile completion.
    typealias ProgressCallback = (
        _ completedTiles: Int,
        _ totalTiles: Int,
        _ bytesDownloaded: Int64,
        _ isComplete: Bool,
        _ error: String?
    ) -> Void

    /// Region metadata persisted in UserDefaults.
    struct RegionMeta: Codable {
        let regionId: String
        let west: Double
        let south: Double
        let east: Double
        let north: Double
        let minZoom: Int
        let maxZoom: Int
        let tileCount: Int
        let createdAt: Date
    }

    // MARK: - Tile URL Templates

    private enum TileSource: CaseIterable {
        case vector
        case satellite
        case terrain

        func url(z: Int, x: Int, y: Int, token: String) -> String {
            switch self {
            case .vector:
                return "https://api.mapbox.com/v4/mapbox.mapbox-streets-v8/\(z)/\(x)/\(y).vector.pbf?access_token=\(token)"
            case .satellite:
                return "https://api.mapbox.com/v4/mapbox.satellite/\(z)/\(x)/\(y)@2x.webp?access_token=\(token)"
            case .terrain:
                return "https://api.mapbox.com/v4/mapbox.terrain-rgb/\(z)/\(x)/\(y).pngraw?access_token=\(token)"
            }
        }
    }

    // MARK: - Style Resource URLs

    /// Build URLs for style JSON, sprites, and glyph ranges (matching Android).
    private static func buildResourceURLs(styleName: String?, token: String) -> [String] {
        guard let styleName = styleName, !styleName.isEmpty else { return [] }

        var urls: [String] = []

        // Style JSON
        urls.append("https://api.mapbox.com/styles/v1/mapbox/\(styleName)?access_token=\(token)")
        // Sprite JSON + PNG
        urls.append("https://api.mapbox.com/styles/v1/mapbox/\(styleName)/sprite@2x.json?access_token=\(token)")
        urls.append("https://api.mapbox.com/styles/v1/mapbox/\(styleName)/sprite@2x.png?access_token=\(token)")

        // Common glyph ranges for Latin + Arabic text
        let fontStacks = [
            "DIN Pro Regular,Arial Unicode MS Regular",
            "DIN Pro Medium,Arial Unicode MS Regular",
            "DIN Pro Bold,Arial Unicode MS Bold"
        ]
        let glyphRanges = [
            "0-255", "256-511", "512-767", "768-1023",
            "8192-8447", "8448-8703", "65024-65279"
        ]
        for fontStack in fontStacks {
            for range in glyphRanges {
                urls.append("https://api.mapbox.com/fonts/v1/mapbox/\(fontStack)/\(range).pbf?access_token=\(token)")
            }
        }

        return urls
    }

    // MARK: - Properties

    private let cacheManager: TileCacheManager

    /// Tracks cancelled region IDs.
    private var cancelledRegions: Set<String> = []
    private let lock = NSLock()

    /// URLSession with limited concurrency.
    private let session: URLSession

    // MARK: - Init

    init(cacheManager: TileCacheManager) {
        self.cacheManager = cacheManager

        let config = URLSessionConfiguration.default
        config.httpMaximumConnectionsPerHost = 4
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300

        let queue = OperationQueue()
        queue.maxConcurrentOperationCount = 4
        queue.name = "ae.trustsky.spotlight.tiledownload"

        session = URLSession(configuration: config, delegate: nil, delegateQueue: queue)
    }

    // MARK: - Download

    /// Start downloading tiles for a region.
    ///
    /// - Parameters:
    ///   - regionId: Unique identifier for the download region.
    ///   - bbox: (west, south, east, north) in degrees.
    ///   - minZoom: Minimum zoom level (inclusive).
    ///   - maxZoom: Maximum zoom level (inclusive).
    ///   - mapboxToken: Mapbox access token.
    ///   - sources: Which tile sources to download. Defaults to all.
    ///   - progress: Callback invoked on every tile completion.
    func downloadRegion(
        regionId: String,
        bbox: (west: Double, south: Double, east: Double, north: Double),
        minZoom: Int,
        maxZoom: Int,
        mapboxToken: String,
        sources: [String]? = nil,
        styleName: String? = nil,
        progress: @escaping ProgressCallback
    ) {
        // Mark as active
        lock.lock()
        cancelledRegions.remove(regionId)
        lock.unlock()

        // Determine which sources to download
        let activeSources: [TileSource]
        if let sourceNames = sources {
            activeSources = sourceNames.compactMap { name -> TileSource? in
                switch name.lowercased() {
                case "vector": return .vector
                case "satellite": return .satellite
                case "terrain": return .terrain
                default: return nil
                }
            }
        } else {
            activeSources = TileSource.allCases
        }

        // Build the list of tile URLs
        var tileURLs: [(url: String, regionPrefix: String)] = []
        for z in minZoom...maxZoom {
            let tileRange = Self.tileRange(bbox: bbox, zoom: z)
            for x in tileRange.xMin...tileRange.xMax {
                for y in tileRange.yMin...tileRange.yMax {
                    for source in activeSources {
                        let url = source.url(z: z, x: x, y: y, token: mapboxToken)
                        tileURLs.append((url: url, regionPrefix: regionId))
                    }
                }
            }
        }

        // Add style, sprite, and glyph resource URLs
        for url in Self.buildResourceURLs(styleName: styleName, token: mapboxToken) {
            tileURLs.append((url: url, regionPrefix: regionId))
        }

        let totalTiles = tileURLs.count
        if totalTiles == 0 {
            progress(0, 0, 0, true, nil)
            return
        }

        // Save region metadata
        let meta = RegionMeta(
            regionId: regionId,
            west: bbox.west, south: bbox.south,
            east: bbox.east, north: bbox.north,
            minZoom: minZoom, maxZoom: maxZoom,
            tileCount: totalTiles,
            createdAt: Date()
        )
        Self.saveRegionMeta(meta)

        // Dispatch downloads
        var completed = 0
        var totalBytes: Int64 = 0
        var lastError: String?
        let statsLock = NSLock()

        let group = DispatchGroup()

        for tile in tileURLs {
            // Check cancellation
            lock.lock()
            let cancelled = cancelledRegions.contains(regionId)
            lock.unlock()
            if cancelled {
                progress(completed, totalTiles, totalBytes, true, "cancelled")
                return
            }

            // Skip tiles already cached
            if cacheManager.get(url: tile.url) != nil {
                statsLock.lock()
                completed += 1
                let c = completed
                let b = totalBytes
                let done = c == totalTiles
                statsLock.unlock()
                progress(c, totalTiles, b, done, nil)
                continue
            }

            group.enter()

            guard let fetchURL = URL(string: tile.url) else {
                group.leave()
                continue
            }

            session.dataTask(with: fetchURL) { [weak self] data, response, error in
                defer { group.leave() }
                guard let self = self else { return }

                // Check cancellation
                self.lock.lock()
                let cancelled = self.cancelledRegions.contains(regionId)
                self.lock.unlock()
                if cancelled { return }

                if let error = error {
                    statsLock.lock()
                    completed += 1
                    lastError = error.localizedDescription
                    let c = completed
                    let b = totalBytes
                    let done = c == totalTiles
                    let e = lastError
                    statsLock.unlock()
                    progress(c, totalTiles, b, done, e)
                    return
                }

                if let data = data,
                   let httpResponse = response as? HTTPURLResponse,
                   (200...299).contains(httpResponse.statusCode) {
                    self.cacheManager.put(url: tile.url, data: data)

                    statsLock.lock()
                    completed += 1
                    totalBytes += Int64(data.count)
                    let c = completed
                    let b = totalBytes
                    let done = c == totalTiles
                    statsLock.unlock()
                    progress(c, totalTiles, b, done, nil)
                } else {
                    statsLock.lock()
                    completed += 1
                    lastError = "HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)"
                    let c = completed
                    let b = totalBytes
                    let done = c == totalTiles
                    let e = lastError
                    statsLock.unlock()
                    progress(c, totalTiles, b, done, e)
                }
            }.resume()
        }

        // Fire a final completion event when every task finishes
        group.notify(queue: .main) {
            statsLock.lock()
            let c = completed
            let b = totalBytes
            let e = lastError
            statsLock.unlock()
            if c < totalTiles {
                progress(c, totalTiles, b, true, e)
            }
        }
    }

    // MARK: - Cancellation

    func cancelDownload(regionId: String) {
        lock.lock()
        cancelledRegions.insert(regionId)
        lock.unlock()
    }

    // MARK: - Slippy Map Math

    private struct TileRange {
        let xMin: Int
        let xMax: Int
        let yMin: Int
        let yMax: Int
    }

    /// Convert a bounding box at a given zoom to tile x/y ranges.
    private static func tileRange(
        bbox: (west: Double, south: Double, east: Double, north: Double),
        zoom: Int
    ) -> TileRange {
        let xMin = lonToTileX(lon: bbox.west, zoom: zoom)
        let xMax = lonToTileX(lon: bbox.east, zoom: zoom)
        let yMin = latToTileY(lat: bbox.north, zoom: zoom) // north → smaller y
        let yMax = latToTileY(lat: bbox.south, zoom: zoom)

        return TileRange(xMin: min(xMin, xMax),
                         xMax: max(xMin, xMax),
                         yMin: min(yMin, yMax),
                         yMax: max(yMin, yMax))
    }

    private static func lonToTileX(lon: Double, zoom: Int) -> Int {
        return Int(floor((lon + 180.0) / 360.0 * pow(2.0, Double(zoom))))
    }

    private static func latToTileY(lat: Double, zoom: Int) -> Int {
        let latRad = lat * .pi / 180.0

        return Int(floor(
            (1.0 - log(tan(latRad) + 1.0 / cos(latRad)) / .pi) / 2.0 * pow(2.0, Double(zoom))
        ))
    }

    // MARK: - Region Metadata Persistence

    private static let metaKey = "ae.trustsky.spotlight.offlinetiles.regions"

    static func saveRegionMeta(_ meta: RegionMeta) {
        var all = loadAllRegionMeta()
        all[meta.regionId] = meta
        if let data = try? JSONEncoder().encode(all) {
            UserDefaults.standard.set(data, forKey: metaKey)
        }
    }

    static func loadAllRegionMeta() -> [String: RegionMeta] {
        guard let data = UserDefaults.standard.data(forKey: metaKey),
              let decoded = try? JSONDecoder().decode([String: RegionMeta].self, from: data) else {
            return [:]
        }

        return decoded
    }

    static func removeRegionMeta(regionId: String) {
        var all = loadAllRegionMeta()
        all.removeValue(forKey: regionId)
        if let data = try? JSONEncoder().encode(all) {
            UserDefaults.standard.set(data, forKey: metaKey)
        }
    }
}

File: android/app/src/main/java/ae/trustsky/spotlight/plugins/TileDownloadManager.java
package ae.trustsky.spotlight.plugins;

import android.util.Log;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Manages bulk downloading of map tiles for offline use. Calculates tile coordinates from a
 * bounding box and zoom range, fetches tiles concurrently, and stores them in the TileCacheManager.
 */
public class TileDownloadManager {

  private static final String TAG = "TileDownloadManager";
  private static final int THREAD_COUNT = 4;
  private static final int CONNECT_TIMEOUT_MS = 15_000;
  private static final int READ_TIMEOUT_MS = 30_000;

  // Tile URL templates — token is appended as a query parameter
  private static final String VECTOR_TEMPLATE =
      "https://api.mapbox.com/v4/mapbox.mapbox-streets-v8/%d/%d/%d.vector.pbf?access_token=%s";
  private static final String SATELLITE_TEMPLATE =
      "https://api.mapbox.com/v4/mapbox.satellite/%d/%d/%d@2x.webp?access_token=%s";
  private static final String TERRAIN_TEMPLATE =
      "https://api.mapbox.com/v4/mapbox.terrain-rgb/%d/%d/%d.pngraw?access_token=%s";
  private static final String STYLE_TEMPLATE =
      "https://api.mapbox.com/styles/v1/mapbox/%s?access_token=%s";
  private static final String GLYPHS_TEMPLATE =
      "https://api.mapbox.com/fonts/v1/mapbox/%s/%s.pbf?access_token=%s";
  private static final String SPRITE_JSON_TEMPLATE =
      "https://api.mapbox.com/styles/v1/mapbox/%s/sprite@2x.json?access_token=%s";
  private static final String SPRITE_PNG_TEMPLATE =
      "https://api.mapbox.com/styles/v1/mapbox/%s/sprite@2x.png?access_token=%s";

  private final TileCacheManager cacheManager;
  private final ExecutorService executor;
  private final ConcurrentHashMap<String, AtomicBoolean> cancelFlags = new ConcurrentHashMap<>();

  public TileDownloadManager(TileCacheManager cacheManager) {
    this.cacheManager = cacheManager;
    this.executor = Executors.newFixedThreadPool(THREAD_COUNT);
  }

  /** Listener for download progress and completion events. */
  public interface DownloadListener {
    void onProgress(String regionId, int completed, int total, float percentage);

    void onComplete(String regionId, int totalTiles, boolean cancelled);

    void onError(String regionId, String error);
  }

  /** Tile layer types that can be downloaded. */
  public enum TileLayer {
    VECTOR,
    SATELLITE,
    TERRAIN
  }

  /** Groups download parameters into a single object. */
  public static class DownloadRequest {
    public final String regionId;
    public final double[] bbox;
    public final int minZoom;
    public final int maxZoom;
    public final List<TileLayer> layers;
    public final String styleName;
    public final String token;

    public DownloadRequest(
        String regionId,
        double[] bbox,
        int minZoom,
        int maxZoom,
        List<TileLayer> layers,
        String styleName,
        String token) {
      this.regionId = regionId;
      this.bbox = bbox;
      this.minZoom = minZoom;
      this.maxZoom = maxZoom;
      this.layers = layers;
      this.styleName = styleName;
      this.token = token;
    }
  }

  /**
   * Download tiles for a region defined by a bounding box and zoom range.
   *
   * @param request Download parameters
   * @param listener Progress callback listener
   */
  public void downloadRegion(DownloadRequest request, DownloadListener listener) {
    if (request.bbox == null || request.bbox.length != 4) {
      if (listener != null) {
        listener.onError(
            request.regionId, "Invalid bounding box: must be [west, south, east, north]");
      }
      return;
    }

    List<String> urls = buildAllUrls(request);
    int totalTiles = urls.size();

    Log.d(TAG, "Starting download for region " + request.regionId + ": " + totalTiles + " resources, zoom " + request.minZoom + "-" + request.maxZoom);

    submitDownloadTasks(request.regionId, urls, totalTiles, listener);
  }

  /**
   * Download tiles for a region (convenience overload with individual parameters).
   */
  public void downloadRegion(
      String regionId,
      double[] bbox,
      int minZoom,
      int maxZoom,
      List<TileLayer> layers,
      String styleName,
      String token,
      DownloadListener listener) {
    downloadRegion(
        new DownloadRequest(regionId, bbox, minZoom, maxZoom, layers, styleName, token), listener);
  }

  /** Cancel an in-progress download. */
  public void cancel(String regionId) {
    AtomicBoolean flag = cancelFlags.get(regionId);
    if (flag != null) {
      flag.set(true);
      cancelFlags.remove(regionId);
      Log.d(TAG, "Cancelled download for region: " + regionId);
    }
  }

  /** Shut down the executor service. */
  public void shutdown() {
    executor.shutdownNow();
  }

  /** Estimate the number of tiles for a region (for UI previews). */
  public static int estimateTileCount(double[] bbox, int minZoom, int maxZoom, int layerCount) {
    if (bbox == null || bbox.length != 4) return 0;

    double west = bbox[0];
    double south = bbox[1];
    double east = bbox[2];
    double north = bbox[3];
    int total = 0;

    for (int z = minZoom; z <= maxZoom; z++) {
      int minX = lonToTileX(west, z);
      int maxX = lonToTileX(east, z);
      int minY = latToTileY(north, z);
      int maxY = latToTileY(south, z);

      int tilesAtZoom = (maxX - minX + 1) * (maxY - minY + 1);
      total += tilesAtZoom;
    }

    return total * layerCount;
  }

  // --- Private helpers ---

  /** Build the complete list of URLs to download (tiles + style resources). */
  private List<String> buildAllUrls(DownloadRequest request) {
    double west = request.bbox[0];
    double south = request.bbox[1];
    double east = request.bbox[2];
    double north = request.bbox[3];

    List<TileCoord> tileCoords = calculateTileCoords(west, south, east, north, request.minZoom, request.maxZoom);
    List<String> urls = buildTileUrls(tileCoords, request.layers, request.token);
    urls.addAll(buildResourceUrls(request.styleName, request.token));
    return urls;
  }

  /** Calculate all tile coordinates for the given bounds and zoom range. */
  private static List<TileCoord> calculateTileCoords(
      double west, double south, double east, double north, int minZoom, int maxZoom) {
    List<TileCoord> coords = new ArrayList<>();
    for (int z = minZoom; z <= maxZoom; z++) {
      int minX = lonToTileX(west, z);
      int maxX = lonToTileX(east, z);
      int minY = latToTileY(north, z);
      int maxY = latToTileY(south, z);

      for (int x = minX; x <= maxX; x++) {
        for (int y = minY; y <= maxY; y++) {
          coords.add(new TileCoord(z, x, y));
        }
      }
    }
    return coords;
  }

  /** Build tile URLs for all coordinates and layers. */
  private static List<String> buildTileUrls(
      List<TileCoord> tileCoords, List<TileLayer> layers, String token) {
    List<String> urls = new ArrayList<>();
    for (TileCoord coord : tileCoords) {
      for (TileLayer layer : layers) {
        urls.add(tileUrlForLayer(layer, coord, token));
      }
    }
    return urls;
  }

  /** Get the URL template for a specific tile layer. */
  private static String tileUrlForLayer(TileLayer layer, TileCoord coord, String token) {
    switch (layer) {
      case VECTOR:
        return String.format(VECTOR_TEMPLATE, coord.z, coord.x, coord.y, token);
      case SATELLITE:
        return String.format(SATELLITE_TEMPLATE, coord.z, coord.x, coord.y, token);
      case TERRAIN:
        return String.format(TERRAIN_TEMPLATE, coord.z, coord.x, coord.y, token);
      default:
        return String.format(VECTOR_TEMPLATE, coord.z, coord.x, coord.y, token);
    }
  }

  /** Build style, sprite, and glyph resource URLs. */
  private static List<String> buildResourceUrls(String styleName, String token) {
    List<String> urls = new ArrayList<>();
    if (styleName == null || styleName.isEmpty()) {
      return urls;
    }

    urls.add(String.format(STYLE_TEMPLATE, styleName, token));
    urls.add(String.format(SPRITE_JSON_TEMPLATE, styleName, token));
    urls.add(String.format(SPRITE_PNG_TEMPLATE, styleName, token));

    // Common glyph ranges for typical Latin + Arabic text
    String[] fontStacks = {
      "DIN Pro Regular,Arial Unicode MS Regular",
      "DIN Pro Medium,Arial Unicode MS Regular",
      "DIN Pro Bold,Arial Unicode MS Bold"
    };
    String[] glyphRanges = {
      "0-255", "256-511", "512-767", "768-1023",
      "8192-8447", "8448-8703", "65024-65279"
    };
    for (String fontStack : fontStacks) {
      for (String range : glyphRanges) {
        urls.add(String.format(GLYPHS_TEMPLATE, fontStack, range, token));
      }
    }
    return urls;
  }

  /** Submit all download tasks to the executor and track progress. */
  private void submitDownloadTasks(
      String regionId, List<String> urls, int totalTiles, DownloadListener listener) {
    AtomicBoolean cancelFlag = new AtomicBoolean(false);
    cancelFlags.put(regionId, cancelFlag);

    AtomicInteger completed = new AtomicInteger(0);

    for (String url : urls) {
      executor.submit(
          () -> {
            if (cancelFlag.get()) return;

            try {
              if (!cacheManager.has(url)) {
                byte[] data = fetchTile(url);
                if (data.length > 0 && !cancelFlag.get()) {
                  cacheManager.put(url, data);
                }
              }
            } catch (IOException e) {
              Log.w(TAG, "Failed to download: " + url, e);
            }

            int done = completed.incrementAndGet();
            if (listener != null && !cancelFlag.get()) {
              float pct = (float) done / totalTiles * 100f;
              listener.onProgress(regionId, done, totalTiles, pct);

              if (done >= totalTiles) {
                cancelFlags.remove(regionId);
                listener.onComplete(regionId, totalTiles, false);
              }
            }
          });
    }
  }

  // --- Slippy map tile coordinate math ---

  /**
   * Convert longitude to tile X coordinate at a given zoom level. x = floor((lon + 180) / 360 *
   * 2^z)
   */
  static int lonToTileX(double lon, int zoom) {
    int n = 1 << zoom; // 2^z
    int x = (int) Math.floor((lon + 180.0) / 360.0 * n);
    return Math.max(0, Math.min(n - 1, x));
  }

  /**
   * Convert latitude to tile Y coordinate at a given zoom level. y = floor((1 - ln(tan(lat_rad) +
   * sec(lat_rad)) / pi) / 2 * 2^z)
   */
  static int latToTileY(double lat, int zoom) {
    int n = 1 << zoom; // 2^z
    double latRad = Math.toRadians(lat);
    int y =
        (int)
            Math.floor(
                (1.0 - Math.log(Math.tan(latRad) + 1.0 / Math.cos(latRad)) / Math.PI) / 2.0 * n);
    return Math.max(0, Math.min(n - 1, y));
  }

  /** Fetch a tile from the network. Returns empty array on non-200 response. */
  private byte[] fetchTile(String urlString) throws IOException {
    HttpURLConnection connection =
        (HttpURLConnection) URI.create(urlString).toURL().openConnection();
    try {
      connection.setConnectTimeout(CONNECT_TIMEOUT_MS);
      connection.setReadTimeout(READ_TIMEOUT_MS);
      connection.setRequestMethod("GET");
      connection.setRequestProperty("User-Agent", "TrustSky-Spotlight-Android/1.0");

      int responseCode = connection.getResponseCode();
      if (responseCode != 200) {
        Log.w(TAG, "HTTP " + responseCode + " for: " + urlString);
        return new byte[0];
      }

      try (InputStream inputStream = connection.getInputStream();
          ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
        byte[] buffer = new byte[8192];
        int bytesRead;
        while ((bytesRead = inputStream.read(buffer)) != -1) {
          baos.write(buffer, 0, bytesRead);
        }
        return baos.toByteArray();
      }
    } finally {
      connection.disconnect();
    }
  }

  /** Internal tile coordinate holder. */
  private static class TileCoord {
    final int z;
    final int x;
    final int y;

    TileCoord(int z, int x, int y) {
      this.z = z;
      this.x = x;
      this.y = y;
    }
  }
}
